apis/v1alpha2/udproute_types.go

```diff
@@ -41,17 +41,29 @@ type UDPRoute struct {
 
 // UDPRouteSpec defines the desired state of UDPRoute.
 type UDPRouteSpec struct {
+	// ParentRefs references the resources that can attach to this Route. The
+	// only kind of parent resource with "Core" support is Gateway. This API may
+	// be extended in the future to support additional kinds of parent resources
+	// such as one of the route kinds. It is invalid to reference an identical
+	// parent more than once. It is valid to reference multiple distinct
+	// sections within the same parent resource, such as 2 Listeners within a
+	// Gateway.
+	//
+	// It is possible to separately reference multiple distinct objects that may
+	// be collapsed by an implementation. For example, some implementations may
+	// choose to merge compatible Gateway Listeners together. If that is the
+	// case, the list of routes attached to those resources should also be
+	// merged.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	ParentRefs []ParentRef `json:"parentRefs,omitempty"`
```

Comment:
My memory's foggy here, but I thought there was some kind of guidance/recommendation to avoid this pattern? Can't find it offhand, will look around to see if I can find any relevant conventions/etc. If not I'll embed like you suggested because it would simplify a lot here.

Reply:
I've been thinking about this some more and I'm concerned about the overhead it would add for implementers. I think this would be more worthwhile if we were sharing more than one field, but since this is limited to `ParentRefs`, it may be simpler to leave this as is. This also enables us to add some variation or nuance per Route, such as specifying the protocols of listeners a given Route type may be attached to or allowing more/less parents for a given Route type.

Reply:
> My memory's foggy here, but I thought there was some kind of guidance/recommendation to avoid this pattern? Can't find it offhand, will look around to see if I can find any relevant conventions/etc.

Tim had a comment saying that embedding is fine if we are absolutely sure that all fields of the struct being embedded will always be applicable. 

> I've been thinking about this some more and I'm concerned about the overhead it would add for implementers

Could you elaborate on the overhead here?

If I have a CustomRoute, now I all I've to do is embed CommonSpec struct and I'm all set as far as the machinery goes.  Same for Spec section, I embed RouteStatus and I'm all set.

> . I think this would be more worthwhile if we were sharing more than one field, but since this is limited to ParentRefs, it may be simpler to leave this as is. 

Even with one field, we have two benefits:
- Custom routes become simpler as demonstrated above
- Avoid doc comment copy pasta in each route type

> This also enables us to add some variation or nuance per Route, such as specifying the protocols of listeners a given Route type may be attached to or allowing more/less parents for a given Route type.

The first one would be an addition to CommonSpec. The second one will require a custom ParentRef.

I can see advantages of incorporating this here. This is not a blocker so please feel free to reject/accept.

Reply:
AFAICT one benefit of embedding would be there is just one place to set the `MaxItems` constraint.

Reply:
> Could you elaborate on the overhead here?

It's not about accessing the fields, it's if you're using the struct creation shortcut, like people often do in testing. In that case, each embed has to be specified exactly, which also means that if we're doing multiple layers of embeds, it can get a bit unwieldy.

https://play.golang.org/p/xbRASdvapbl- illustrates here. You can still get or set the embedded-embedded fields, but creating a whole object needs extra cruft.

Obviously, there's a tradeoff between this and the ease of use that comes from only having one place to update the embedded type, but we've got similar problems in a few places, so I think we should try and keep away from >1 layer of embedding where we can. (I came across then when looking at deduping the various ObjectReference and BackendRef types). The other tradeoff is, of course, that we need to ensure we don't have unnecessary YAML indendation required (which isn't the case here).

In this case, I think the embedding is good, because of the advantages for people who want to make a custom Route type.

Reply:
This all makes sense, I've gone ahead and introduced an embedded `CommonRouteSpec` struct.